amivapi/auth/oauth.py

```diff
@@ -0,0 +1,219 @@
+
+    if response_type != 'token':
+        abort(422, "response_type is not supported.")
+
+    db = current_app.data.driver.db['oauthclients']
+    client = db.find_one({'client_id': client_id})
+
+    if not client:
+        abort(422, "Unknown client_id")
+
+    if not redirect_uri:
+        redirect_uri = client['redirect_uri']
+
+    if not redirect_uri.startswith(client['redirect_uri']):
+        abort(422, "Redirect URI is not whitelisted!")
+
+    return redirect_uri
+
+
+@oauth_blueprint.route('/oauth/login', methods=['POST'])
+def oauth_login():
+    """Endpoint to receive user input from the login form."""
+    response_type = request.form.get('response_type')
+    client_id = request.form.get('client_id')
+    redirect_uri = request.form.get('redirect_uri')
+    scope = request.form.get('scope')
+    state = request.form.get('state')
+
+    # Check this is a proper oauth client
+    redirect_uri = validate_oauth_authorization_request(
+        response_type, client_id, redirect_uri)
+
+    # Check we have login data
+    token = request.cookies.get('token')
+    try:
+        if token:
+            # In this case the user was logged in already and just accepted
+            # the request.
+            authenticate_token(token)
+            if g.current_user is None:
+                # Session ended since we served the login page. Back to login.
+                # This is caught by the except below.
+                abort(401)
+
+        else:
+            resp = post_internal(
+                'sessions',
+                {
+                    'username': request.form.get('username'),
+                    'password': request.form.get('password')
+                }
+            )[0]
+            if 'token' not in resp:
+                abort(401, "post_internal to sessions failed: %s" % resp)
+            token = resp['token']
+
+    except Unauthorized as e:
+        # Login failed. Redirect back to the login page.
+        error_msg = ("Login failed! If you think there is an error, please "
+                     "contact AMIV with the exact time of your login.")
+        current_app.logger.info("Login failed with error: %s" % e)
+        login_url = url_for('.oauth_authorize',
+                            response_type=response_type,
+                            client_id=client_id,
+                            redirect_url=redirect_uri,
+                            scope=scope,
+                            state=state,
+                            error_msg=error_msg)
+        response = current_app.make_response(redirect(login_url))
+        response.set_cookie('token', '', expires=0)
+        return response
+
+    # We have a valid token! Let's bring the user back to the oauth client.
+    redirect_uri = append_url_params(redirect_uri, {
+        'access_token': token,
+        'token_type': 'bearer',
+        'scope': 'amiv',
+        'state': state
+    })
+    response = current_app.make_response(redirect(redirect_uri))
```

Comment:
I would move this line down a little to make clear that we need to call `make_response` in order to attach the cookie. If one is not experienced with flask, this might be confusing otherwise.